File: optimal.java
import java.util.*;

public class OptimalPageReplacement {

    public static void optimalPageReplacement(int[] pages, int frameSize) {
        List<Integer> frames = new ArrayList<>();
        int pageFaults = 0;

        for (int i = 0; i < pages.length; i++) {
            int page = pages[i];

            if (!frames.contains(page)) {
                pageFaults++;
                if (frames.size() < frameSize) {
                    frames.add(page);
                } else {
                    // Find the page to replace
                    int indexToReplace = -1;
                    int farthestUse = -1;

                    for (int j = 0; j < frames.size(); j++) {
                        int f = frames.get(j);
                        int nextUse = Integer.MAX_VALUE;

                        // Check future uses
                        for (int k = i + 1; k < pages.length; k++) {
                            if (pages[k] == f) {
                                nextUse = k;
                                break;
                            }
                        }

                        if (nextUse > farthestUse) {
                            farthestUse = nextUse;
                            indexToReplace = j;
                        }
                    }
                    frames.set(indexToReplace, page);
                }
            }
            System.out.println("Frames after accessing " + page + ": " + frames);
        }

        System.out.println("\nTotal Page Faults: " + pageFaults);
    }

    public static void main(String[] args) {
        int[] pages = {2, 3, 2, 1, 5, 2, 4, 5, 3, 2, 5, 2};
        int frameSize = 3;

        optimalPageReplacement(pages, frameSize);
    }
}

File: Python/optimal.py
# Optimal Page Replacement Algorithm in Python

def optimal_page_replacement(pages, frame_size):
    frames = []
    page_faults = 0

    for i in range(len(pages)):
        page = pages[i]

        if page not in frames:
            page_faults += 1
            if len(frames) < frame_size:
                frames.append(page)
            else:
                # Find the page to replace
                future_uses = []
                for f in frames:
                    if f in pages[i+1:]:
                        future_uses.append(pages[i+1:].index(f))
                    else:
                        # If page is not used again, replace it immediately
                        future_uses.append(float('inf'))
                
                # Replace the page that is used farthest in the future
                index_to_replace = future_uses.index(max(future_uses))
                frames[index_to_replace] = page
        print(f"Frames after accessing {page}: {frames}")

    print(f"\nTotal Page Faults: {page_faults}")

# Sample input
pages = [2, 3, 2, 1, 5, 2, 4, 5, 3, 2, 5, 2]
frame_size = 3

optimal_page_replacement(pages, frame_size)
